haystack/components/rankers/metadata_grouper.py

```diff
@@ -0,0 +1,190 @@
+        self.sort_docs_by = sort_docs_by
+        self.subgroup_by = subgroup_by
+
+    def _buck_into_groups(self, documents: List[Document]) -> Tuple[Dict[str, List], List, List]:
+        """
+        Go through all documents and bucket them based on the 'group_by' value.
+
+        If no 'group_by' value is present in the document, the document is added to a list of documents without a group.
+
+        :param documents:
+        :returns:
+            A tuple with the following elements:
+            - document_groups: A dictionary with the 'group_by' values as keys and the documents as values.
+            - no_group: A list of documents without a 'group_by' value.
+            - ordered_keys: A list of 'group_by' values in the order they were encountered.
+        """
+
+        document_groups: Dict[str, List] = defaultdict(list)
+        ordered_keys: List = []
+        no_group: List = []
+
+        for document in documents:
+            if self.group_by in document.meta:
+                document_groups[str(document.meta[self.group_by])].append(document)
+                if str(document.meta[self.group_by]) not in ordered_keys:
+                    ordered_keys.append(str(document.meta[self.group_by]))
+            else:
+                no_group.append(document)
+
+        return document_groups, no_group, ordered_keys
+
+    def _buck_into_subgroups(self, document_groups: Dict[str, List]) -> Tuple[Dict[str, Dict], List]:
+        """
+        Buckets the documents within the groups based on the 'subgroup_by' value.
+
+        If no 'subgroup_by' value is present in the document, the document is added to a subgroup with the key
+        'no_subgroup'.
+
+        :param document_groups: A dictionary with the 'group_by' values as keys and the documents as values.
+        :returns:
+            A tuple with the following elements:
+            - document_subgroups: A dictionary with the 'subgroup_by' values as keys and the documents as values.
+            - subgroup_ordered_keys: A list of 'subgroup_by' values in the order they were encountered
+        """
+        subgroup_ordered_keys = []
+        document_subgroups: Dict[str, Dict] = defaultdict(lambda: defaultdict(list))
+        if self.subgroup_by:
+            for key, value in document_groups.items():
+                for doc in value:
+                    if self.subgroup_by in doc.meta:
+                        document_subgroups[key][str(doc.meta[self.subgroup_by])].append(doc)
+                        if str(doc.meta[self.subgroup_by]) not in subgroup_ordered_keys:
+                            subgroup_ordered_keys.append(str(doc.meta[self.subgroup_by]))
+                    else:
+                        document_subgroups[key]["no_subgroup"].append(doc)
+                        if "no_subgroup" not in subgroup_ordered_keys:
+                            subgroup_ordered_keys.append("no_subgroup")
+        return document_subgroups, subgroup_ordered_keys
```

Comment:
this could become similar to this

```python
    def _create_subgroups(self, groups: Dict[str, List[Document]]) -> Dict[str, Dict[str, List[Document]]]:
        if not self.subgroup_by:
            return {}

        subgroups: Dict[str, Dict[str, List[Document]]] = defaultdict(lambda: defaultdict(list))
        
        for group_key, docs in groups.items():
            for doc in docs:
                subgroup_value = self._get_meta_value(doc, self.subgroup_by)
                subgroups[group_key][subgroup_value].append(doc)
                
        return subgroups
```

Reply:
here you also miss the handling documents whose metadata doesn't match the subgroup key provided at init time and also doesn't collect the values for those that have the metadata subgroup key